File: pkg/worker/worker.go
package worker

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/HughNian/nmid/pkg/alert"
	"github.com/HughNian/nmid/pkg/logger"
	"github.com/HughNian/nmid/pkg/model"
	"github.com/HughNian/nmid/pkg/trace"
	"github.com/HughNian/nmid/pkg/utils"
	"github.com/SkyAPM/go2sky"
	clientv3 "go.etcd.io/etcd/client/v3"
)

//rpc tcp worker

type Worker struct {
	sync.Mutex

	WorkerId   string
	WorkerName string

	Agents   []*Agent
	Funcs    map[string]*Function
	FuncsNum int
	Resps    chan *Response
	stop     chan struct{}

	Reporter go2sky.Reporter
	Tracer   *go2sky.Tracer

	ready    bool
	running  bool
	useTrace bool
}

func NewWorker() *Worker {
	wor := &Worker{
		Agents:   make([]*Agent, 0),
		Funcs:    make(map[string]*Function),
		FuncsNum: 0,
		Resps:    make(chan *Response, model.QUEUE_SIZE),
		ready:    false,
		running:  false,
		useTrace: false,
		stop:     make(chan struct{}),
	}

	return wor
}

func (w *Worker) SetWorkerId(wid string) *Worker {
	if len(wid) == 0 {
		w.WorkerId = utils.GetId()
	} else {
		w.WorkerId = wid
	}
	return w
}

func (w *Worker) SetWorkerName(wname string) *Worker {
	if len(wname) == 0 {
		w.WorkerName = utils.GetId()
	} else {
		w.WorkerName = wname
	}

	return w
}

func (w *Worker) WithTrace(reporterUrl string) *Worker {
	w.useTrace = true
	w.Reporter, w.Tracer = trace.NewReporter(reporterUrl, w.WorkerName)
	return w
}

func (w *Worker) AddServer(net, addr string) (err error) {
	var agent *Agent = NewAgent(net, addr, w)

	if agent == nil {
		return fmt.Errorf("agent nil")
	}
	w.Agents = append(w.Agents, agent)

	return nil
}

func (w *Worker) Register(config EtcdConfig) {
	etcdcli := EtcdClient(config)
	if etcdcli == nil {
		return
	}
	defer etcdcli.Close()

	var nmidAddrs []string
	for _, val := range w.Agents {
		nmidAddrs = append(nmidAddrs, val.addr)
	}

	workerIns := make(map[string][]string)
	for _, val := range w.Funcs {
		workerIns[val.FuncName] = nmidAddrs
	}

	jret, err := json.Marshal(workerIns)
	if err != nil {
		return
	}
	workerVal := string(jret)
	workerKey := fmt.Sprintf("%s%s", model.EtcdBaseKey, w.WorkerId)

	kv := clientv3.NewKV(etcdcli)
	_, err = kv.Put(context.TODO(), workerKey, workerVal)
	if err != nil {
		logger.Error("register err", err)
	}
}

func (w *Worker) AddFunction(funcName string, jobFunc JobFunc) (err error) {
	w.Lock()
	if _, ok := w.Funcs[funcName]; ok {
		return fmt.Errorf("function already exist")
	}

	w.Funcs[funcName] = NewFunction(jobFunc, funcName)
	w.FuncsNum++
	w.Unlock()

	if w.running {
		go w.FuncBroadcast(funcName, model.PDT_W_ADD_FUNC)
	}

	return nil
}

func (w *Worker) DelFunction(funcName string) (err error) {
	w.Lock()
	if _, ok := w.Funcs[funcName]; !ok {
		return fmt.Errorf("function not exist")
	}

	delete(w.Funcs, funcName)
	w.FuncsNum--
	w.Unlock()

	if w.running {
		go w.FuncBroadcast(funcName, model.PDT_W_DEL_FUNC)
	}

	return nil
}

func (w *Worker) GetFunction(funcName string) (function *Function, err error) {
	if len(w.Funcs) == 0 || w.FuncsNum == 0 {
		return nil, fmt.Errorf("worker have no funcs")
	}

	w.Lock()
	f, ok := w.Funcs[funcName]
	w.Unlock()

	if f == nil || !ok {
		return nil, fmt.Errorf("not found")
	}

	if f.FuncName != funcName {
		return nil, fmt.Errorf("not found")
	}

	function = f

	return function, nil
}

func (w *Worker) DoFunction(resp *Response) (err error) {
	if resp.DataType == model.PDT_S_GET_DATA {
		//use trace
		if w.useTrace {
			//set entry span
			resp.SetEntrySpan()
		}

		funcName := resp.Handle
		if function, err := w.GetFunction(funcName); err != nil {
			return err
		} else if function != nil {
			if function.FuncName != funcName {
				return fmt.Errorf("funcname error")
			}
			if resp.ParamsLen == 0 {
				return fmt.Errorf("params error")
			}

			var ret []byte
			if ret, err = function.Func(resp); err == nil {
				resp.Agent.Req.HandleLen = resp.HandleLen
				resp.Agent.Req.Handle = resp.Handle
				resp.Agent.Req.ParamsLen = resp.ParamsLen
				resp.Agent.Req.Params = resp.Params
				resp.Agent.Req.JobIdLen = resp.JobIdLen
				resp.Agent.Req.JobId = resp.JobId

				resp.Agent.Lock()
				resp.Agent.Req.RetPack(ret)
				resp.Agent.Write()
				resp.Agent.Unlock()
			}
		}
	}

	return nil
}

func (w *Worker) FuncBroadcast(funcName string, flag int) {
	for _, a := range w.Agents {
		switch flag {
		case model.PDT_W_ADD_FUNC:
			a.Req.AddFunctionPack(funcName)
		case model.PDT_W_DEL_FUNC:
			a.Req.DelFunctionPack(funcName)
		default:
			a.Req.AddFunctionPack(funcName)
		}
		a.Write()
	}
}

func (w *Worker) WorkerReady() (err error) {
	if len(w.Agents) == 0 {
		return errors.New("none active agents")
	}
	if w.FuncsNum == 0 || len(w.Funcs) == 0 {
		return errors.New("none functions")
	}

	for _, a := range w.Agents {
		if err = a.Connect(); err != nil {
			return err
		}
	}

	for fn := range w.Funcs {
		w.FuncBroadcast(fn, model.PDT_W_ADD_FUNC)
	}

	w.Lock()
	w.ready = true
	w.Unlock()

	return nil
}

func (w *Worker) WorkerDo() {
	if !w.ready {
		err := w.WorkerReady()
		if err != nil {
			logger.Fatal(err)
		}
	}

	w.Lock()
	w.running = true
	w.Unlock()

	//nmid server timeout/close and reconnect
	go func() {
		for w.running {
			for _, a := range w.Agents {
				select {
				case <-w.stop:
					return
				default:
					diffTime := utils.GetNowSecond() - a.lastTime
					if diffTime > model.NMID_SERVER_TIMEOUT {
						logger.Error(fmt.Sprintf("nmid server connect timeout or close, nmid server@address:%s, @worker:%s", a.addr, w.WorkerName))
						alert.SendMarkDownAtAll(alert.DERROR, "nmid server error", fmt.Sprintf("nmid server connect timeout or close, nmid server@address:%s, @worker:%s", a.addr, w.WorkerName))

						w.WorkerReConnect(a)
					}
				}
			}

			time.Sleep(5 * time.Second)
		}
	}()

	//send heartbeat ping & grab job
	go func() {
		for w.running {
			select {
			case <-w.stop:
				return
			default:
				duration := model.DEFAULTHEARTBEATTIME
				timer := time.NewTimer(duration)
				<-timer.C

				for _, a := range w.Agents {
					a.HeartBeatPing()
					a.Grab()
				}
			}
		}
	}()

	for resp := range w.Resps {
		switch resp.DataType {
		case model.PDT_TOSLEEP:
			time.Sleep(time.Duration(2) * time.Second)
			go resp.Agent.Wakeup()

			//fallthrough
		case model.PDT_S_GET_DATA:
			if err := w.DoFunction(resp); err != nil {
				logger.Error(err)
			}
			//fallthrough
		case model.PDT_NO_JOB:
			go resp.Agent.Grab()
		case model.PDT_S_HEARTBEAT_PONG:
			resp.Agent.lastTime = utils.GetNowSecond()
		case model.PDT_WAKEUPED:
		default:
			go resp.Agent.Grab()
		}
	}
}

func (w *Worker) WorkerClose() {
	if w.running {
		logger.Info("worker close")

		for fn := range w.Funcs {
			w.FuncBroadcast(fn, model.PDT_W_DEL_FUNC)
		}

		for _, a := range w.Agents {
			a.Close()
		}

		w.stop <- struct{}{}
		w.running = false
		close(w.Resps)

		if w.useTrace {
			w.Reporter.Close()
		}
	}
}

func (w *Worker) WorkerReConnect(a *Agent) {
	//del function msg
	for fname := range w.Funcs {
		a.DelOldFuncMsg(fname)
	}
	//disconnect old
	if a.conn != nil {
		a.conn.Close()
	}

	//reconnect
	a.ReConnect()
	//resend add function msg
	for fname := range w.Funcs {
		a.ReAddFuncMsg(fname)
	}
}

File: sdk/php/src/Worker/Worker.php
<?php
namespace Nmid\Worker;

use Nmid\Constants;
use Nmid\RetStruct;

// nmid worker. Mirrors node lib/worker/worker.js.
//
// Holds one or more Agents (connections to nmid servers) and a set of
// registered functions. The main loop (workerDo) uses stream_select to wait
// for jobs, dispatches them to the registered callable, and returns the ret.
// A heartbeat ping is sent on a fixed interval; a stale agent (no PONG within
// NMID_SERVER_TIMEOUT) is reconnected and re-registered.
class Worker
{
    public string $workerId = '';
    public string $workerName = '';
    /** @var Agent[] */
    public array $agents = [];
    /** @var callable[] name => callable(Response):string */
    public array $funcs = [];
    public int $funcsNum = 0;
    public bool $ready = false;
    public bool $running = false;

    public function setWorkerId(string $wid = ''): self
    {
        $this->workerId = $wid !== '' ? $wid : $this->genId();
        return $this;
    }

    public function setWorkerName(string $wname = ''): self
    {
        $this->workerName = $wname !== '' ? $wname : $this->genId();
        return $this;
    }

    private function genId(): string
    {
        return substr((string)(microtime(true) * 1000), 0, 10) . '-' . mt_rand();
    }

    // network: 'tcp'; addr: "host:port"
    public function addServer(string $network, string $addr): void
    {
        $this->agents[] = new Agent($network, $addr, $this);
    }

    public function addFunction(string $funcName, callable $jobFunc): void
    {
        if (isset($this->funcs[$funcName])) {
            throw new \RuntimeException("function {$funcName} already exist");
        }
        $this->funcs[$funcName] = $jobFunc;
        $this->funcsNum++;
    }

    public function delFunction(string $funcName): void
    {
        if (!isset($this->funcs[$funcName])) {
            throw new \RuntimeException("function {$funcName} not exist");
        }
        unset($this->funcs[$funcName]);
        $this->funcsNum--;
    }

    // Connect all agents, set worker name and register all functions, waiting
    // for PDT_OK after each registration packet (avoids the "have no job do"
    // race for early client calls).
    public function workerReady(): ?\Throwable
    {
        if (empty($this->agents)) {
            return new \RuntimeException('none active agents');
        }
        if ($this->funcsNum === 0 || empty($this->funcs)) {
            return new \RuntimeException('none funcs');
        }

        if ($this->workerName === '') {
            $this->setWorkerName();
        }

        foreach ($this->agents as $agent) {
            try {
                $agent->connect();
            } catch (\Throwable $e) {
                return $e;
            }

            $agent->req->setWorkerName($this->workerName);
            $agent->write();
            $agent->expectOk();

            foreach (array_keys($this->funcs) as $fname) {
                $agent->req->addFunctionPack($fname);
                $agent->write();
                $agent->expectOk();
            }
        }

        $this->ready = true;
        return null;
    }

    // Dispatch an incoming decoded response to the right handler.
    public function handleResp(Response $resp): void
    {
        $dt = $resp->dataType;

        if ($dt === Constants::PDT_OK) {
            // Handled by expectOk() during registration.
            return;
        }

        if ($dt === Constants::PDT_TOSLEEP) {
            // Server asks us to sleep briefly then wake up. PHP is synchronous,
            // so just send wakeup immediately.
            if ($resp->agent !== null) {
                $resp->agent->wakeup();
            }
            return;
        }

        if ($dt === Constants::PDT_S_GET_DATA) {
            $err = $this->doFunction($resp);
            if ($err !== null) {
                fwrite(STDERR, "error: " . $err->getMessage() . "\n");
            }
            return;
        }

        if ($dt === Constants::PDT_S_HEARTBEAT_PONG) {
            if ($resp->agent !== null) {
                $resp->agent->lastTime = (int)(microtime(true) * 1000);
            }
            return;
        }

        // PDT_NO_JOB / PDT_WAKEUPED: nothing to do.
    }

    // Execute the registered function for a job and return its ret to the server.
    public function doFunction(Response $resp): ?\Throwable
    {
        if (!isset($this->funcs[$resp->handle])) {
            return new \RuntimeException("function {$resp->handle} not found");
        }
        if ($resp->paramsLen === 0) {
            return new \RuntimeException('params error');
        }

        $agent = $resp->agent;
        try {
            $ret = ($this->funcs[$resp->handle])($resp);
            if (!is_string($ret)) {
                return new \RuntimeException('job function must return a string');
            }
            $agent->req->returnDataPack($resp, $ret);
            $agent->write();
            return null;
        } catch (\Throwable $e) {
            return $e;
        }
    }

    public function heartBeat(): void
    {
        foreach ($this->agents as $agent) {
            if ($agent->conn !== null) {
                $agent->heartBeatPing();
            }
        }
    }

    public function workerReConnect(Agent $agent): void
    {
        $agent->close();
        fwrite(STDERR, "warning: reconnecting agent\n");
        try {
            $agent->connect();
        } catch (\Throwable $e) {
            fwrite(STDERR, "reconnect failed: " . $e->getMessage() . "\n");
            return;
        }

        // Re-register and wait for PDT_OK, same as workerReady.
        $agent->req->setWorkerName($this->workerName);
        $agent->write();
        $agent->expectOk();
        foreach (array_keys($this->funcs) as $fname) {
            $agent->req->addFunctionPack($fname);
            $agent->write();
            $agent->expectOk();
        }
    }

    // Main loop: connect + register, then run heartbeat + timeout + job loop.
    public function workerDo(): void
    {
        if (!$this->ready) {
            $e = $this->workerReady();
            if ($e !== null) {
                throw $e;
            }
        }

        $this->running = true;
        $this->heartBeat();
        $lastHeartbeat = microtime(true);

        while ($this->running) {
            $read = [];
            $agentByStream = [];
            foreach ($this->agents as $agent) {
                if ($agent->conn !== null) {
                    $read[] = $agent->conn;
                    $agentByStream[(int)$agent->conn] = $agent;
                }
            }
            if (empty($read)) {
                break;
            }

            $write = null;
            $except = null;
            // 1s select so we can still fire heartbeats / timeout checks.
            $n = @stream_select($read, $write, $except, 1);
            $now = microtime(true);

            if ($n === false) {
                // Interrupted (e.g. by a signal); loop again.
                continue;
            }

            if ($n > 0) {
                foreach ($read as $stream) {
                    $agent = $agentByStream[(int)$stream] ?? null;
                    if ($agent === null) {
                        continue;
                    }
                    if (!$agent->onData()) {
                        // Connection closed — reconnect.
                        $this->workerReConnect($agent);
                    }
                }
            }

            // Heartbeat on a fixed interval.
            if ($now - $lastHeartbeat >= Constants::DEFAULT_HEARTBEAT_TIME) {
                $this->heartBeat();
                $lastHeartbeat = $now;
            }

            // Reconnect stale agents (no PONG within NMID_SERVER_TIMEOUT).
            foreach ($this->agents as $agent) {
                if ($agent->conn !== null
                    && ($now * 1000 - $agent->lastTime) > Constants::NMID_SERVER_TIMEOUT) {
                    $this->workerReConnect($agent);
                }
            }
        }
    }

    public function workerClose(): void
    {
        $this->running = false;
        foreach ($this->agents as $agent) {
            $agent->close();
        }
    }

    // Helper for example code to build a standard RetStruct.
    public static function buildRet(int $code, string $msg, string $data = ''): string
    {
        return RetStruct::build($code, $msg, $data);
    }
}
